usersession/userd/privileged_desktop_launcher.go

```diff
@@ -0,0 +1,312 @@
+// We're not required to diagnose multiple files matching the desktop file ID.
+func findDesktopFile(baseDir string, splitFileId []string) (string, error) {
+	desktopFile := filepath.Join(baseDir, strings.Join(splitFileId, "-"))
+
+	exists, isReg, _ := regularFileExists(desktopFile)
+	if exists && isReg {
+		return desktopFile, nil
+	}
+
+	// Iterate through the potential subdirectories formed by the first i elements of the desktop file ID.
+	// Maybe this is overkill: At the time of writing, the only use is in desktopFileIDToFilename() and there
+	// we're only checking dirs.SnapDesktopFilesDir (not all entries in $XDG_DATA_DIRS) and we know that snapd
+	// does not create subdirectories in that location.
+	for i := 1; i != len(splitFileId); i++ {
+		prefix := strings.Join(splitFileId[:i], "-")
+		// Don't treat empty or "." components as directory
+		// prefixes.  The ".." case is already filtered out by
+		// the isValidDesktopFileID regexp.
+		if prefix == "" || prefix == "." {
+			continue
+		}
+		desktopFile, err := findDesktopFile(filepath.Join(baseDir, prefix), splitFileId[i:])
+		if err == nil {
+			return desktopFile, nil
+		}
+	}
+
+	return "", fmt.Errorf("could not find desktop file")
+}
+
+// isValidDesktopFileID is based on the "File naming" section of the
+// Desktop Entry Specification, without the restriction on components
+// not starting with a digit (which desktop files created by snapd may
+// not satisfy).
+var isValidDesktopFileID = regexp.MustCompile(`^[A-Za-z0-9-_]+(\.[A-Za-z0-9-_]+)*.desktop$`).MatchString
+
+// desktopFileIDToFilename determines the path associated with a desktop file ID.
+func desktopFileIDToFilename(desktopFileID string) (string, error) {
+	if !isValidDesktopFileID(desktopFileID) {
+		return "", fmt.Errorf("cannot find desktop file for %q", desktopFileID)
+	}
+
+	splitDesktopID := strings.Split(desktopFileID, "-")
+	for _, baseDir := range desktopFileSearchPath() {
+		if desktopFile, err := findDesktopFile(baseDir, splitDesktopID); err == nil {
+			return desktopFile, nil
+		}
+	}
+
+	return "", fmt.Errorf("cannot find desktop file for %q", desktopFileID)
+}
+
+// verifyDesktopFileLocation checks the desktop file location:
+// we only consider desktop files in dirs.SnapDesktopFilesDir
+func verifyDesktopFileLocation(desktopFile string) error {
+	if filepath.Clean(desktopFile) != desktopFile {
+		return fmt.Errorf("desktop file has unclean path: %q", desktopFile)
+	}
+
+	if !strings.HasPrefix(desktopFile, dirs.SnapDesktopFilesDir+"/") {
+		// We currently only support launching snap applications from desktop files in
+		// /var/lib/snapd/desktop/applications and these desktop files are written by snapd and
+		// considered safe for userd to process. If other directories are added in the future,
+		// verifyDesktopFileLocation() and parseExecCommand() may need to be updated.
+		return fmt.Errorf("internal error: only launching snap applications from %s is supported", dirs.SnapDesktopFilesDir)
```

Comment:
given the new approach to search this is not an internal error anymore, the prefix should be stripped